### PR TITLE
feat: implement ledger onboarding flows

### PR DESCRIPTION
**Pull request by laumair**
## Summary

This PR sets up the onboarding flow for ledger profiles

## Relevant Issues

https://github.com/iotaledger/firefly/issues/3948

### Platforms

> Please select any platforms where your changes have been tested.

- __Desktop__
  - [x] MacOS
  - [ ] Linux
  - [ ] Windows
- __Mobile__
  - [ ] iOS
  - [ ] Android

### Instructions

**How to setup ledger:** https://hackmd.io/@laumair/S1-Uyrc05

## Checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or modified tests that prove my changes work as intended
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have verified that my latest changes pass CI workflows for testing and linting
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: packages/shared/components/modals/ProfileActions.svelte [hunk 1]
```diff
@@ -61,31 +58,14 @@
...
-            //         ledgerConnectionText = `${la.name} ${text}`
-            //     })
-            //     .catch((err) => {
-            //         ledgerDeviceState.set(LedgerDeviceState.NotDetected)
-            //         console.error(err)
-            //     })
-        } else {
-            ledgerConnectionText = text
-        }
+        ledgerConnectionText = text
```
Comment (maxwellmattryan) [inline packages/shared/components/modals/ProfileActions.svelte:68]: We can leave the `text` variable declaration if we need to use it later, but otherwise could just set it directly with `ledgerConnectionText = localize(`views.dashboard.profileModal.hardware.statuses.${$ledgerDeviceStatus.connectionState}`)`

File: packages/shared/components/modals/ProfileActions.svelte [hunk 2]
```diff
@@ -2,11 +2,10 @@
     import { fade } from 'svelte/transition'
     import { Button, DeveloperIndicatorPill, HR, Icon, Modal, Text, Toggle } from 'shared/components'
     import { localize } from '@core/i18n'
-    import { LocaleArguments } from '@core/i18n/types'
-    import { ledgerDeviceState } from 'shared/lib/ledger'
+    import { LedgerConnectionState } from '@lib/typings/ledger'
+    import { ledgerDeviceStatus, getLedgerDeviceStatus } from 'shared/lib/ledger'
```
Comment (maxwellmattryan) [inline packages/shared/components/modals/ProfileActions.svelte:6]: We've adjusted the codebase quite a bit from V1 - now we're trying to slowly whittle down all code that's not in `lib/core`, `lib/contexts` nor `lib/auxiliary`. Perhaps we can refactor the code in `shared/lib/ledger.ts` to a new `core/ledger` module?

File: packages/shared/lib/ledger.ts [hunk 3]
```diff
@@ -1,67 +1,62 @@
 import { get, writable } from 'svelte/store'
```
Comment (maxwellmattryan) [inline packages/shared/lib/ledger.ts:1]: I think we should refactor this file into a new `lib/core/ledger` module :raised_hands:

File: packages/shared/lib/typings/ledger.ts [hunk 4]
```diff
@@ -1,37 +1,21 @@
...
     NotDetected = 'notDetected',
     OtherConnected = 'otherConnected',
+    MnemonicMismatch = 'mnemonicMismatch',
 }
 
 export enum LedgerAppName {
     BOLOS = 'BOLOS',
-    IOTA = 'IOTA',
+    SHIMMER = 'Shimmer',
     IOTALegacy = 'IOTA Legacy',
```
Comment (maxwellmattryan) [inline packages/shared/lib/typings/ledger.ts:19]: We can remove this enum variant since V2 won't support the legacy migration :+1: (same goes for any migration-related code).

File: packages/shared/locales/en.json [hunk 5]
```diff
@@ -191,7 +191,7 @@
         },
         "ledgerInstallationGuide": {
             "title": "Have you installed the necessary Ledger apps?",
-            "body1": "Before you proceed, you must find and install both the {legacy} app and the new IOTA app through Ledger Live.",
+            "body1": "Before you proceed, you must use Ledger Live to find and install Shimmer app.",
```
Comment (maxwellmattryan) [inline packages/shared/locales/en.json:194]: We can use a `{protocol}` value to make this locale a little more useful

File: packages/shared/locales/en.json [hunk 6]
```diff
@@ -446,7 +446,7 @@
                 "hardware": {
                     "title": "Hardware Device",
                     "statuses": {
-                        "appNotOpen": "IOTA app not open",
+                        "appNotOpen": "Shimmer app not open",
```
Comment (maxwellmattryan) [inline packages/shared/locales/en.json:449]: Same as above

File: packages/shared/locales/en.json [hunk 7]
```diff
@@ -701,7 +701,7 @@
             "bodyFirefly": "Firefly is disabled until it is complete. Token migration will still be possible afterwards."
         },
         "ledgerNotConnected": {
-            "connect": "Open the new IOTA app on your Ledger device to continue",
+            "connect": "Open the Shimmer app on your Ledger device to continue",
```
Comment (maxwellmattryan) [inline packages/shared/locales/en.json:704]: Same as above

File: packages/shared/lib/typings/ledger.ts [hunk 8]
```diff
@@ -1,37 +1,21 @@
...
-export enum LedgerDeviceState {
+export enum LedgerConnectionState {
     AppNotOpen = 'appNotOpen',
     Connected = 'connected',
-    LegacyConnected = 'legacyConnected',
     Locked = 'locked',
-    MnemonicMismatch = 'mnemonicMismatch',
     NotDetected = 'notDetected',
     OtherConnected = 'otherConnected',
+    MnemonicMismatch = 'mnemonicMismatch',
```
Comment (maxwellmattryan) [inline packages/shared/lib/typings/ledger.ts:13]: Small detail but since the other variants are in past tense, it might read a bit better to have `MnemonicMismatched` instead?

File: packages/shared/routes/onboarding/views/ledger-setup/views/ConnectView.svelte [hunk 9]
```diff
@@ -47,28 +42,22 @@
     }
 
     function handleContinueClick(): void {
```
Comment (maxwellmattryan) [inline packages/shared/routes/onboarding/views/ledger-setup/views/ConnectView.svelte:44]: Small detail, but I've been renaming all of these click handler functions to `onContinueClick` or `onBackClick` to help readability with the `OnboardingLayout` component props.

File: packages/shared/routes/onboarding/views/ledger-setup/views/InstallationGuideView.svelte [hunk 10]
```diff
@@ -26,22 +26,15 @@
...
         <Text type="p" secondary classes="mb-5">
-            {localize('views.ledgerInstallationGuide.body1', { values: { legacy: LedgerAppName.IOTALegacy } })}
+            {localize('views.ledgerInstallationGuide.body1')}
         </Text>
         <Text type="p" secondary classes="mb-5">{localize('views.ledgerInstallationGuide.body2')}</Text>
         <div class="flex flex-row flex-nowrap items-center justify-center space-x-4 text-center mt-28">
             <div class="flex flex-col flex-wrap space-y-2">
                 <div class="bg-blue-400 rounded-2xl w-20 h-20 flex justify-center items-center">
-                    <Icon icon="ledger-app-legacy" width="32" height="32" classes="text-white" />
+                    <Icon icon="shimmer" width="32" height="32" classes="text-white" />
```
Comment (maxwellmattryan) [inline packages/shared/routes/onboarding/views/ledger-setup/views/InstallationGuideView.svelte:35]: We should double check if we only want to show this view during Ledger onboarding for only Shimmer and not IOTA. I assumed we will show for both, but since only the Shimmer app is new maybe it makes sense to leave it?

File: packages/shared/lib/typings/ledger.ts (hunk 4 above)
Comment (maxwellmattryan) [inline packages/shared/lib/typings/ledger.ts:19]: I think I might've made this enum before our coding conventions were set - it should be:
```typescript
export enum LedgerAppName {
    Bolos = 'bolos',
    Iota = 'iota',
    Shimmer = 'shimmer',
}
```

This might affect locale stuff wrt the Ledger connection status, but not entirely sure.

**Review by maxwellmattryan — CHANGES_REQUESTED** [review]
Code is looking pretty good! Have requested some changes.

Thanks :pray:

File: packages/shared/lib/core/account/actions/syncSelectedAccount.ts [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import { localize } from '@core/i18n'
 import { isLedgerProfile } from '@core/profile'
-import { displayNotificationForLedgerProfile } from '@lib/ledger'
+import { displayNotificationForLedgerProfile } from '@lib/core/ledger'
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/account/actions/syncSelectedAccount.ts:3]: We can shorten this to `@core/ledger`

File: packages/shared/lib/core/account/actions/tryCreateAdditionalAccount.ts [hunk 12]
```diff
@@ -1,6 +1,6 @@
 import { localize } from '@core/i18n'
 import { isLedgerProfile } from '@core/profile'
-import { displayNotificationForLedgerProfile } from '@lib/ledger'
+import { displayNotificationForLedgerProfile } from '@lib/core/ledger'
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/account/actions/tryCreateAdditionalAccount.ts:3]: This can also be shortened.

File: packages/shared/lib/core/account/actions/tryEditSelectedAccountMetadata.ts [hunk 13]
```diff
@@ -1,6 +1,6 @@
 import { localize } from '@core/i18n'
 import { isLedgerProfile, updateActiveAccountMetadata } from '@core/profile'
-import { displayNotificationForLedgerProfile } from '@lib/ledger'
+import { displayNotificationForLedgerProfile } from '@lib/core/ledger'
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/account/actions/tryEditSelectedAccountMetadata.ts:3]: This can also be shortened.

File: packages/shared/lib/core/ledger/enums/ledger-app-name.enum.ts [hunk 14]
```diff
@@ -0,0 +1,4 @@
+export enum LedgerAppName {
+    BOLOS = 'BOLOS',
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/ledger/enums/ledger-app-name.enum.ts:2]: Forgot to mention last time - the enum variant names should be in `PascalCase` and the corresponding values in `camelCase`. I can't remember enough to know if the camelCase values will break anything, like locale usage or something.

File: packages/shared/lib/core/ledger/interfaces/ledger-extended-status.interface.ts [hunk 15]
```diff
@@ -0,0 +1,6 @@
+import { LedgerStatus } from './ledger-status.interface'
+import { LedgerConnectionState } from './ledger-connection-state.interface'
+
+export interface LedgerExtendedStatus extends LedgerStatus {
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/ledger/interfaces/ledger-extended-status.interface.ts:4]: Is there a reason why we're keeping the status and extended status separately?

File: packages/shared/lib/core/ledger/utils/determineLedgerDeviceState.ts [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+        return LedgerConnectionState.Locked
+    } else {
+        switch (app?.name) {
+            default:
+                if (connected) {
+                    /**
+                     * NOTE: "BOLOS" is the name of the Ledger operating system and is
+                     * sometimes registered as an app.
+                     */
+                    return app?.name && app?.name !== LedgerAppName.BOLOS
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/ledger/utils/determineLedgerDeviceState.ts:16]: Ah, probably changing enum values will break this logic?

File: packages/shared/lib/core/profile-manager/utils/buildProfileManagerOptionsFromProfileData.ts [hunk 17]
```diff
@@ -1,20 +1,27 @@
-import { getStorageDirectoryOfProfile, IPersistedProfile } from '@core/profile'
+import { getStorageDirectoryOfProfile, IPersistedProfile, ProfileType } from '@core/profile'
 import { ProfileManagerOptions } from '@core/profile-manager'
 import { COIN_TYPE, getDefaultClientOptions } from '@core/network'
+import { USE_LEDGER_SIMULATOR } from '@lib/core/ledger'
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/profile-manager/utils/buildProfileManagerOptionsFromProfileData.ts:4]: We can shorten this one too :+1:

File: packages/shared/lib/core/profile/stores/login-progress.store.ts [hunk 18]
```diff
@@ -4,10 +4,10 @@ import { ILoginProgress } from '../interfaces'
 
 export const loginProgress = writable<ILoginProgress>({ stepCount: 0, stepMessage: LOGIN_STEPS[0] })
 
-export function incrementLoginProgress(): void {
+export function incrementLoginProgress(incrementBy = 1): void {
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/profile/stores/login-progress.store.ts:7]: Sorry to be pedantic :sweat_smile:, can we add a `number` type for `incrementBy`?

File: packages/shared/lib/core/profile/actions/active-profile/logout.ts [hunk 19]
```diff
@@ -1,5 +1,5 @@
 import { resetRouters } from '@core/router'
-import { stopPollingLedgerStatus } from '@lib/ledger'
+import { stopPollingLedgerStatus } from '@lib/core/ledger'
```
Comment (maxwellmattryan) [inline packages/shared/lib/core/profile/actions/active-profile/logout.ts:2]: This can be shortened :+1:

File: packages/shared/locales/en.json [hunk 20]
```diff
@@ -191,7 +191,7 @@
         },
         "ledgerInstallationGuide": {
             "title": "Have you installed the necessary Ledger apps?",
-            "body1": "Before you proceed, you must find and install both the {legacy} app and the new IOTA app through Ledger Live.",
+            "body1": "Before you proceed, you must use Ledger Live to find and install {protocol} app.",
```
Comment (maxwellmattryan) [inline packages/shared/locales/en.json:194]: Can we add a "the" before protocol, e.g. `...to find and install the {protocol} app`?

File: packages/shared/locales/en.json [hunk 21]
```diff
@@ -459,9 +459,8 @@
                 "hardware": {
                     "title": "Hardware Device",
                     "statuses": {
-                        "appNotOpen": "IOTA app not open",
+                        "appNotOpen": "Shimmer app not open",
```
Comment (maxwellmattryan) [inline packages/shared/locales/en.json:462]: Should we use a `{protocol}` ICU message arg here?

File: packages/shared/routes/dashboard/settings/views/advanced/MigrateLedgerIndex.svelte [hunk 22]
```diff
@@ -1,11 +1,10 @@
 <script lang="typescript">
     import { Button, Text } from 'shared/components'
     import { localize } from '@core/i18n'
-    import { navigateToNewIndexMigration } from 'shared/lib/ledger'
 </script>
 
 <Text type="h4" classes="mb-3">{localize('views.settings.migrateLedgerIndex.title')}</Text>
 <Text type="p" secondary classes="mb-5">{localize('views.settings.migrateLedgerIndex.description')}</Text>
-<Button medium inlineStyle="min-width: 156px;" onClick={navigateToNewIndexMigration}>
+<Button medium inlineStyle="min-width: 156px;" onClick={() => {}}>
```
Comment (maxwellmattryan) [inline packages/shared/routes/dashboard/settings/views/advanced/MigrateLedgerIndex.svelte:8]: Can we remove the `onClick` attribute since it's an empty function?

**Review by maxwellmattryan — CHANGES_REQUESTED** [review]
Looking good! Left a few minor comments for requested changes :pray:

File: packages/shared/lib/core/ledger/utils/determineLedgerDeviceState.ts (hunk 16 above)
Comment (laumair) [reply packages/shared/lib/core/ledger/utils/determineLedgerDeviceState.ts:16]: Yep, I think we can keep it as is?

File: packages/shared/lib/core/ledger/interfaces/ledger-extended-status.interface.ts (hunk 15 above)
Comment (laumair) [reply packages/shared/lib/core/ledger/interfaces/ledger-extended-status.interface.ts:4]: `LedgerStatus` is what wallet-rs returns us and we add the connection status for our purposes.

File: packages/shared/lib/core/profile/stores/login-progress.store.ts (hunk 18 above)
Comment (laumair) [reply packages/shared/lib/core/profile/stores/login-progress.store.ts:7]: It can be inferred by the default value, but I will add it.

**Comment by maxwellmattryan** [comment]
I ran into the following problem when testing:

- Restoring from Ledger device, got this when I clicked on "Finish setup". It went to the loading screen for a while, then brought me back to the congratulations view with this error (seemed to fail on "Generating accounts" step):
    ```
    login.ts:102 Error: Account with ID null cannot be found!
        at setSelectedAccount (setSelectedAccount.ts:12)
        at login (login.ts:93)
    ```

    The strange thing is that clicking "Finish setup" again went to the loading screen again and properly went through the loading bar to the dashboard. It seems to have create a `NaN` profile too :thinking: 
    
    Screenshot from 2022-08-18 16-34-47

**Comment by laumair** [comment]
> I ran into the following problem when testing:
> 
> * Restoring from Ledger device, got this when I clicked on "Finish setup". It went to the loading screen for a while, then brought me back to the congratulations view with this error (seemed to fail on "Generating accounts" step):
>   ```
>   login.ts:102 Error: Account with ID null cannot be found!
>       at setSelectedAccount (setSelectedAccount.ts:12)
>       at login (login.ts:93)
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   The strange thing is that clicking "Finish setup" again went to the loading screen again and properly went through the loading bar to the dashboard. It seems to have create a `NaN` profile too 🤔
>   Screenshot from 2022-08-18 16-34-47
> 
> Creating a new profile works flawlessly!

It looks like that for me, it gets stuck on `Generating accounts` for ledger profiles. Informed wallet-rs team about it. For now, reducing the account gap limit to 1 for ledger profiles as done in https://github.com/iotaledger/firefly/pull/4221/commits/7fd21083e29d27be94253ecc252156e357efc3e5, works.

File: packages/shared/package.json [hunk 23]
```diff
@@ -35,6 +35,7 @@
         "@rollup/plugin-typescript": "^6.0.0",
         "@tsconfig/svelte": "^1.0.10",
         "@types/jest": "^27.0.1",
+        "@types/node": "^18.7.9",
```
Comment (maxwellmattryan) [inline packages/shared/package.json:38]: I suppose this is needed as a dep and not just a dev dependency?

**Review by maxwellmattryan — COMMENTED** [review]
LGTM! Left one question

File: packages/shared/package.json (hunk 23 above)
Comment (laumair) [reply packages/shared/package.json:38]: Nah, but we only need types for development so it should be a dev dependency.

**Review by maxwellmattryan — APPROVED** [review]
